### PR TITLE
Quit all processes when "apiserver-boot run local" failed

### PR DESCRIPTION
**Pull request by interma**
Fix:
https://github.com/kubernetes-incubator/apiserver-builder/issues/253

Glad to open my first PR, thanks for your review.

### CODE REVIEW

**Comment by damongant** [comment]
I can't review this, but I'd suggest you do the following:

* Create a new `context.Context`
* Create a cancel function from it
* Give all processes the cancel function and context
* In the run functions create a new `<-chan error`, run `cmd.Run` in a new goroutine handing over the channel and sending the return value of that to the channel as soon as it exits
* outside the goroutine select from `context.Done()` and the channel you created. If the your channel case is hit, cancel the context, if the context is canceled by another process exiting (the other case) kill the process and return.
* make sure your context also cancels when the process receives an Interrupt or Kill signal (Ctrl+C in console for instance)

```golang
func cancelWhenSignaled(parent context.Context) context.Context {
	ctx, cancel := context.WithCancel(parent)

	go func() {
		signalChannel := make(chan os.Signal)
		signal.Notify(signalChannel, os.Interrupt, os.Kill)
		<-signalChannel
		cancel()
	}()

	return ctx
}
```

**Comment by interma** [comment]
Thanks @damongant 
I will update later.

**Comment by interma** [comment]
@damongant  
I have refined quit logic as your advice (yes, it more makes sense). Did I catch your idea?

Thanks for your help.

File: cmd/apiserver-boot/boot/run/local.go [hunk 1]
```diff
@@ -178,45 +159,58 @@ func RunApiserver(stopCh chan struct{}) *exec.Cmd {
...
+
+	fmt.Printf("%s\n", strings.Join(cmd.Args, " "))
 	go func() {
-		err := controllerManagerCmd.Run()
+		err := cmd.Run()
 		if err != nil {
-			log.Printf("Failed to run controller-manager %v", err)
-			stopCh <- struct{}{}
+			log.Printf("Failed to run %s, error: %v\n", cmdName, err)
+			stopCh <- err
```
Comment (damongant) [inline cmd/apiserver-boot/boot/run/local.go]: You can save this line and the else and just `stopCh<-err` - it'll be nil in the else case either way.

File: cmd/apiserver-boot/boot/run/local.go [hunk 2]
```diff
@@ -178,45 +159,58 @@ func RunApiserver(stopCh chan struct{}) *exec.Cmd {
...
 	}()
 
-	return controllerManagerCmd
+	select {
+	case <-stopCh:
+		// my command quited
+		cancel()
+	case <-ctx.Done():
+		// other commands quited
+		cmd.Process.Kill()
```
Comment (damongant) [inline cmd/apiserver-boot/boot/run/local.go]: do we need to check if `cmd.Process` is `nil` here? Unfortunate timing with 2 processes exiting could produce a segfault here afaik, but I'm not sure.

**Comment by damongant** [comment]
LGTM overall, didn't have a chance to test yet.

File: cmd/apiserver-boot/boot/run/local.go (hunk 2 above)
Comment (interma) [reply cmd/apiserver-boot/boot/run/local.go]: Yes, check nil is better, I will update them later.

File: cmd/apiserver-boot/boot/run/local.go (hunk 1 above)
Comment (interma) [reply cmd/apiserver-boot/boot/run/local.go]: ok, I will update this.

**Comment by interma** [comment]
@pwittrock  Could you help to review this? 
Thanks.

**Comment by schittler** [comment]
I'd hate for this contribution to go to waste because of inactive maintainers.

wink wink nudge nudge

/remove-lifecycle stale

File: cmd/apiserver-boot/boot/util/util.go [hunk 3]
```diff
@@ -139,3 +142,16 @@ func CheckInstall() {
 			strings.Join(missing, ","))
 	}
 }
+
+func CancelWhenSignaled(parent context.Context) context.Context {
```
Comment (yue9944882) [inline cmd/apiserver-boot/boot/util/util.go]: reusing existing signal handler from here seems better to me : 

[https://github.com/kubernetes/kubernetes/blob/7f23a743e8c23ac6489340bbb34fa6f1d392db9d/staging/src/k8s.io/sample-controller/pkg/signals/signal.go#L29](https://github.com/kubernetes/kubernetes/blob/7f23a743e8c23ac6489340bbb34fa6f1d392db9d/staging/src/k8s.io/sample-controller/pkg/signals/signal.go#L29)

Comment (yue9944882) [reply cmd/apiserver-boot/boot/util/util.go]: @interma any plan to bump the thread?

Comment (schittler) [reply cmd/apiserver-boot/boot/util/util.go]: double-receiving should be a fine addition to this piece of code if that's what you mean, but the context is also used to support cancelling when one of the processes unexpectedly quits.

Comment (yue9944882) [reply cmd/apiserver-boot/boot/util/util.go]: > double-receiving should be a fine addition to this piece of code if that's what you mean

for now we're not catching signal in the code so there's no double-receiving. i mean the pull LGTM overall, but we should reuse the existing code from upstream as much as possible.

**Comment by interma** [comment]
I remember apiserver-builder is deprecated last year, reloads now?
What's the purpose and roadmap now?
Thanks.

**Comment by yue9944882** [comment]
> I remember apiserver-builder is deprecated last year, reloads now?

for now, we don't see a better replacement tooling/SDK for AA server so we're still delivering it. actually there were supposed to be a replacement for AA-builder, we planed and named it `apiserver-runtime`. but out of some reason, it's delayed and we don't have a clear roadmap for it yet. 

> What's the purpose and roadmap now?

generally, we're aiming at converging this project w/ kubebuilder (which is now actively maintained by phil and his team) to share the same framework like `controller-runtime` etc. fwiw the point of this project is reusing the wheels from kube-builder as much as possible and extending features beyond CRD's capability. CRD's going more light-weight and better integrated w/ openapi-schema and AA's going more extensible for customization.

File: cmd/apiserver-boot/boot/util/util.go (hunk 3 above)
Comment (wallrj) [reply cmd/apiserver-boot/boot/util/util.go]: I've attempted to address this final code-review comment in https://github.com/kubernetes-sigs/apiserver-builder-alpha/pull/410

It'd be great to get this fixed in the next release.

Comment (wallrj) [reply cmd/apiserver-boot/boot/util/util.go]: Trouble is that importing  SetupSignalHandler from k8s.io/apiserver causes a clash with the k8s.io/kube-openapi dependency.
So I copied the code across instead.

**Comment by wallrj** [comment]
@interma I forked your branch and addressed the remaining code review feedback in https://github.com/kubernetes-sigs/apiserver-builder-alpha/pull/410 which has now been merged, so this can now be closed. I hope you don't mind.

**Comment by interma** [comment]
@wallrj Ok, no problem, thanks!

**Pull request closed without merging**